tools/mock-okta/index.js

```diff
@@ -149,6 +152,45 @@ function handleKeys(req, res) {
 }
 
 /**
+ * For the /oauth2/default/v1/token request, we need to replace the dynamic state
+ * in the request form-body with a known state
+ * This is to ensure that the right tape is requested in yakbak,
+ * which uses body to hash-request the tape
+ *
+ * In this case, a known state is the state that was generated while recording
+ */
+function handleToken(req, res, next) {
+  textParser(req, res, () => {
+    if (!req.body) return;
+    // Express openid-client sends the dynamic state parameter to token request
+    // We need to replace it with a "known" state, so that we hit the right tape each time
+    req.body = req.body.replace(`state=${store['state']}`, 'state=STATE');
+
+    // Spring back-end sends the client-id and client-secret in the token request.
+    // We use express back-end to record tapes, which sends the state value in token request
+    // During playback, we replace the spring back-end's token request body with state, so that we hit the right tape
+    req.body = req.body.replace(`client_id=${clientConfig.clientId}&client_secret=${clientConfig.clientSecret}`, 
+                                'state=STATE');
+
+    // Massaging the request headers to match the change in body length
+    req.headers['content-length'] = `${req.body.length}`;
+
+    // Spring back-end doesn't use Basic Auth for token requests whereas express back-end does
```

Comment:
Curious if this is an implementation on our part, or Spring Security. Both should be exposed.

Reply:
Our token endpoint supports both, but as we're seeing here some clients use the header rather than the body.

Reply:
AFAIK, it's an implementation in Spring security, not our library.